tests/zfs-tests/cmd/mkholes/mkholes.c

```diff
@@ -0,0 +1,255 @@
+{
+	int	rand_fd;
+	char	*buf;
+
+	buf = umem_alloc(len, UMEM_NOFAIL);
+
+	/*
+	 * Fill the buffer from /dev/urandom to counteract the
+	 * effects of compression.
+	 */
+	if ((rand_fd = open("/dev/urandom", O_RDONLY)) < 0) {
+		perror("open /dev/urandom failed");
+		exit(1);
+	}
+
+	if (read(rand_fd, buf, len) < 0) {
+		perror("read /dev/urandom failed");
+		exit(1);
+	}
+
+	(void) close(rand_fd);
+
+	return (buf);
+}
+
+void
+list_create(list_t *list, size_t size, size_t offset)
+{
+	ASSERT(list);
+	ASSERT(size > 0);
+	ASSERT(size >= offset + sizeof (list_node_t));
+
+	list->list_size = size;
+	list->list_offset = offset;
+	list->list_head.list_next = list->list_head.list_prev =
+	    &list->list_head;
+}
+
+void
+list_insert_tail(list_t *list, void *object)
+{
+	list_node_t *lold = &list->list_head;
+	list_insert_before_node(list, lold, object);
+}
+
+void *
+list_remove_head(list_t *list)
+{
+	list_node_t *head = list->list_head.list_next;
+	if (head == &list->list_head)
+		return (NULL);
+	list_remove_node(head);
+	return (list_object(list, head));
+}
+
+static void
+push_segment(list_t *seg_list, seg_type_t seg_type, char *optarg)
+{
+	char		*off_str, *len_str;
+	static off_t	file_size = 0;
+	off_t		off, len;
+	seg_t		*seg;
+
+	off_str = strtok(optarg, ":");
+	len_str = strtok(NULL, ":");
+
+	if (off_str == NULL || len_str == NULL)
+		usage("Bad offset or length", 1);
+
+	off = strtoull(off_str, NULL, 0);
+	len = strtoull(len_str, NULL, 0);
+
+	if (file_size >= off + len)
+		usage("Ranges must ascend and may not overlap.", 1);
+	file_size = off + len;
+
+	seg = umem_alloc(sizeof (seg_t), UMEM_NOFAIL);
+	seg->seg_type = seg_type;
+	seg->seg_offset = off;
+	seg->seg_len = len;
+
+	list_insert_tail(seg_list, seg);
+}
+
+int
+main(int argc, char *argv[])
+{
+	int	c, fd;
+	char	*fname;
+	list_t	seg_list;
+	seg_t	*seg;
+
+	umem_nofail_callback(no_memory);
+	list_create(&seg_list, sizeof (seg_t), offsetof(seg_t, seg_node));
+
+	while ((c = getopt(argc, argv, "d:h:")) != -1) {
+		switch (c) {
+		case 'd':
+			push_segment(&seg_list, SEG_DATA, optarg);
+			break;
+		case 'h':
+			push_segment(&seg_list, SEG_HOLE, optarg);
+			break;
+		}
+	}
+	argc -= optind;
+	argv += optind;
+
+	if ((fname = argv[0]) == NULL)
+		usage("No filename specified", 1);
+	fname = argv[0];
+
+	if ((fd = open(fname, O_LARGEFILE | O_RDWR | O_CREAT | O_SYNC,
+	    00666)) < 0) {
+		perror("open failed");
+		exit(1);
+	}
+
+	while ((seg = list_remove_head(&seg_list)) != NULL) {
+		char	*buf, *vbuf;
+		off_t	off = seg->seg_offset;
+		off_t	len = seg->seg_len;
+
+		if (seg->seg_type == SEG_HOLE) {
+			off_t bytes_read = 0;
+			ssize_t readlen = 1024 * 1024 * 16;
+
+			if (fallocate(fd, FALLOC_FL_KEEP_SIZE |
+			    FALLOC_FL_PUNCH_HOLE, off, len) != 0) {
+				perror("falloc_fl_punch_hole failed");
+				exit(1);
+			}
+
+			buf = (char *)umem_alloc(readlen, UMEM_NOFAIL);
+			vbuf = (char *)umem_zalloc(readlen, UMEM_NOFAIL);
+			while (bytes_read < len) {
```

Comment:
`fallocate(fd, FALLOC_FL_KEEP_SIZE | FALLOC_FL_PUNCH_HOLE, off, len)` on an empty file won't extend it from `off` of `len` bytes because of `FALLOC_FL_KEEP_SIZE`: `bytes_read` will always be 0 and thus `< len`, this will cause an infinite loop here and is the reason the BB is hitting the 10m limit on `send-c_embedded_blocks`.